[llvm-debuginfo-analyzer] Add support for LLVM IR format.

CarlosAlbertoEnciso opened the pull request:
Add support for the LLVM IR format and be able to generate logical views.

Both textual representation (.ll) and bitcode (.bc) format are supported.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Add support for LLVM IR format.

Add support for the LLVM IR format and be able to generate
logical views.

Both textual representation (.ll) and bitcode (.bc) format
are supported.

Note: This patch requires:

  Add DebugSSAUpdater class to track debug value liveness
  https://github.com/llvm/llvm-project/pull/135349

CarlosAlbertoEnciso commented:
> @CarlosAlbertoEnciso If this needs another pair of eyes, do not hesitate to add me as reviewer. Cheers!

@jalopezg-git Thanks; that would be great.

jmorse commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2348 @@
+  LVElement *Element = constructElement(Context);
+  if (Element) {
+    const DIScope *ParentContext = nullptr;
+    if (const auto *SP = dyn_cast<DISubprogram>(Context)) {
+      // Check for a specific 'Unit'.
+      if (DICompileUnit *CU = SP->getUnit())
+        ParentContext = getMDScope(SP->getDeclaration() ? CU : Context);
```
I think what this is doing is placing function definitions into their namespace/class scope rather than the CU that they get defined in? If so, please comment as such.

jmorse commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2348 @@
+  return NodeIndexType;
+}
+
+// addGlobalName - Add a new global name to the compile unit.
+void LVIRReader::addGlobalName(StringRef Name, LVElement *Element,
+                               const DIScope *Context) {
+  assert(Element && "Invalid logical element.");
+  LLVM_DEBUG({
+    dbgs() << "\n[addGlobalName] DIScope\n";
+    Context->dump();
+  });
+}
```
Is this unimplemented / otherwise unused?

jmorse commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2348 @@
+      DeclArgs = SPDecl->getType()->getTypeArray();
+      DefinitionArgs = SP->getType()->getTypeArray();
+
+      if (DeclArgs.size() && DefinitionArgs.size())
```
Explanatory comment needed -- this is over-riding the declarations type with that of the definitions type? (If so, why?)

jmorse commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2348 @@
+  Aggregate->setIsFinalized();
+
+  dwarf::Tag Tag = Aggregate->getTag();
+  if (Tag == dwarf::DW_TAG_variant_part) {
+  }
```
Unimplemented?

jmorse commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2348 @@
+    AbstractScope->setName(OriginScope->getName());
+    AbstractScope->setType(OriginScope->getType());
+
+    AbstractScope->setCallLineNumber(InlinedAt->getLine());
+    AbstractScope->setCallFilenameIndex(
+        getOrCreateSourceID(InlinedAt->getFile()));
+
+    AbstractScope->setReference(OriginScope);
+    AbstractScope->setHasReferenceAbstract();
```
The object name is `AbstractScope`, but you're setting the inlining-site information on it and referring to the "origin" scope from it. This really sounds like the "Abstract" scope is actually the _inlined_ scope. An inlined subroutine in DWARF is what contains that information:

```
0x0000008a:     DW_TAG_inlined_subroutine
                  DW_AT_abstract_origin (0x00000036 "Max")
                  DW_AT_low_pc  (0x000000000000000b)
                  DW_AT_high_pc (0x0000000000000014)
                  DW_AT_call_file       ("/home/jmorse/test2.c")
                  DW_AT_call_line       (16)
                  DW_AT_call_column     (0x0c)
```
i.e. it has a call line, file index, and reference to another scope.

jmorse reviewed (commented):
Some initial comments; I made it to about 1000 lines into LVIRReader.cpp.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Add support for LLVM IR format.

Add support for the LLVM IR format and be able to generate
logical views.

Both textual representation (.ll) and bitcode (.bc) format
are supported.

Note: This patch requires:

Add DebugSSAUpdater class to track debug value liveness
https://github.com/llvm/llvm-project/pull/135349

Note: Address reviewers comments.

CarlosAlbertoEnciso commented:
Uploaded new patch to address @jmorse comments.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Add support for LLVM IR format.

Fixed compilation issues after latest DebugSSAUpdater rebase.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] IR Reader.

Updated test cases that failed due to an upstream fix related
to the high-pc value being now exclusive.

  https://github.com/llvm/llvm-project/pull/153318

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] IR Reader.

Updated test cases that failed due to the DebugSSAUpdater rebase.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] IR Reader.

Add a missing space between the offset and 'ptr' for a
better layout.

  [004] 9 {Variable} 'CONSTANT' -> 'const INTEGER'
  [006]     {Location}
  [007]       {Entry} bregx 3 ptr %CONSTANT+0

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] IR Reader.

Fix issue reported by Clang format.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] IR Reader.

Fix issue with C++98 error:

  extra ';' outside of a function is incompatible with C++98
  [-Werror,-Wc++98-compat-extra-semi]

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] IR Reader.

Changes required after rebase with upstream to include:

  Remove LVScope::Children container.
  https://github.com/llvm/llvm-project/pull/144750

jalopezg-git reviewed (commented):
Many thanks for this huge contribution, @CarlosAlbertoEnciso :+1:!

I went through all the PR except for `llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp` (which I am going to visit in a second pass next week).

I have attached some comments that I have collected.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+    Context->dump();
+  });
+
+  if (Context && isa<DICompositeType>(Context)) {
+    const DICompositeType *CTy = cast<DICompositeType>(Context);
+    constructType(static_cast<LVScope *>(Element), CTy);
+  } else if (Context && isa<DIDerivedType>(Context)) {
+    const DIDerivedType *DT = cast<DIDerivedType>(Context);
+    constructType(Element, DT);
+  } else if (Context && isa<DISubprogram>(Context)) {
+    const DISubprogram *SP = cast<DISubprogram>(Context);
+    getOrCreateSubprogram(static_cast<LVScope *>(Element), SP);
```
```suggestion
  if (const DICompositeType *CTy = dyn_cast_if_present<DICompositeType>(Context)) {
    constructType(static_cast<LVScope *>(Element), CTy);
  } else if (const DIDerivedType *DT = dyn_cast_if_present<DIDerivedType>(Context)) {
    constructType(Element, DT);
  } else if (const DISubprogram *SP = dyn_cast_if_present<DISubprogram>(Context)) {
    getOrCreateSubprogram(static_cast<LVScope *>(Element), SP);
```

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+          Parameter->setIsArtificial();
+      }
+    } else {
+      // Add an unspecified parameter.
+      Parameter = createElement(dwarf::DW_TAG_unspecified_parameters);
```
`DW_TAG_unspecified_parameters` is used to represent variadic functions.  This looks more like 'unknown' type, but not totally sure.  Should a `DW_TAG_formal_parameter` be created instead, but without setting type information?

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+      Count = CI->getSExtValue();
+    else if (auto *UI =
+                 dyn_cast_if_present<ConstantInt *>(GSR->getUpperBound())) {
+      // Fortran uses 1 as the default lowerbound; other languages use 0.
+      int64_t Lowerbound = (moduleIsInFortran()) ? 1 : 0;
```
It seems like being Fortran or not does not matter, except for the subrange lowerbound.  Should we store the default-array-lowerbound (instead of `LanguageIsFortran`)?

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+    LVScope *Parent = Scope ? Scope : getParentScope(Ty);
+    Parent->addElement(Element);
+
+    if (/*const DIBasicType *BT =*/dyn_cast<DIBasicType>(Ty)) {
```
```suggestion
    if (isa<DIBasicType>(Ty)) {
```

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+    if (N > 1) {
+      const DIType *Ty = Args[N - 1];
+      if (!Ty)
+        AddUnspecifiedParameters = true;
```
Ditto (about `DW_TAG_unspecified_parameters`) -- I am not sure about this one though.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+      BitCodeIR ? BitCodeIR->getMemoryBufferRef() : *TextualIR, Err, Context);
+  if (!M)
+    return createStringError(errc::invalid_argument,
+                             "Could not create IR information: %s",
```
```suggestion
                             "Could not create IR module for: %s",
```

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+
+    Root->addElement(CompileUnit);
+
+    if (options().getAttributeProducer())
+      CompileUnit->setProducer(CU->getProducer());
```
I think we should also set the `CompileUnit` source language, i.e.
```suggestion
    if (options().getAttributeLanguage())
      CompileUnit->setSourceLanguage(LVSourceLanguage{static_cast<llvm::dwarf::SourceLanguage>(CU->getSourceLanguage())});
    if (options().getAttributeProducer())
      CompileUnit->setProducer(CU->getProducer());
```

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+    options().setInternalID();
+}
+
+// During the IR Reader development, traverse all the logical elements
+// to check if they have been properly constructed (finalized).
```
This comment is a bit confusing; maybe you meant
```suggestion
// During the IR-to-logical-view construction, traverse all the logical elements
// to check if they have been properly constructed (finalized).
```
?

jalopezg-git reviewed (commented):
Thank you for this huge work, @CarlosAlbertoEnciso :+1:!  I have finalized the second (and final) part of my review and I have attached some comments below.

I would appreciate though that some expert in the IR DebugInfo representation could take an additional look at `llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp`.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] IR Reader.

Changes required after rebase with upstream to include:

  [llvm][DebugInfo][NFC] Abstract DICompileUnit::SourceLanguage
  to allow alternate DWARF SourceLanguage encoding (#162255)

  https://github.com/llvm/llvm-project/pull/162255

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] IR Reader.

Rust generates multiple #dbg_declare's.

  %self.dbg.spill = alloca [8 x i8], align 8
  store ptr %self, ptr %self.dbg.spill, align 8
    #dbg_declare(ptr %self.dbg.spill, !892, !DIExpression(), !894)
    #dbg_declare(ptr %self.dbg.spill, !895, !DIExpression(), !905)

Change the assertion for a warning and return, otherwise
the IR Reader will not be able to process Rust IR's.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] IR Reader.

Add the ability to print basic information about a logical
element. Mainly is used to print information while dealing
with IR.

It uses the current command line arguments to control what
to print. The information is printed in a single line. In
the minimum case it includes name and type.

CarlosAlbertoEnciso commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
```diff
@@ -0,0 +1,304 @@
+
+  void addMD(const MDNode *MD, LVElement *Element) {
+    if (MDObjects.find(MD) == MDObjects.end())
+      MDObjects.emplace(MD, Element);
```
Good replacement.

CarlosAlbertoEnciso commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+
+    StringRef Name = getMDName(DN);
+    if (!Name.empty())
+      Element->setName(Name);
```
Nice change.

CarlosAlbertoEnciso commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+  return Element;
+}
+
+void LVIRReader::mapFortranLanguage(unsigned DWLang) {
```
Good point. Now we use `LVSourceLanguage`:
```
void LVIRReader::setDefaultLowerBound(LVSourceLanguage *SL) {
  assert(SL && "Invalid language ID.");
  StringRef LanguageName = SL->getName();

  // Fortran uses 1 as the default lowerbound; other languages use 0.
  DefaultLowerBound = LanguageName.contains("fortran") ? 1 : 0;

  LLVM_DEBUG({ dbgs() << "Language Name: " << LanguageName << "\n"; });
}
```
Now it covers `DWARF` and `CodeView`.

CarlosAlbertoEnciso commented on llvm/include/llvm/DebugInfo/LogicalView/Core/LVSupport.h:
```diff
@@ -119,6 +119,19 @@ template <typename T> class LVProperties {
+  std::string String;
+  raw_string_ostream Stream(String);
+  Stream << decValue(Value, Width);
+  return Stream.str();
```
Great point.

CarlosAlbertoEnciso commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```diff
@@ -0,0 +1,2309 @@
+    uint64_t Value = DIExpr->getElement(1);
+    Stream << hexString(Value, 2);
+    Element->setValue(Stream.str());
+  }
```
Nice simplification.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Add support for LLVM IR format.

- Addressed all the reviewers comments.
  * Added test case for multiple compile units.

- Reviewed the inlining functionality issues, found when
  processing IRs produced by the Russ compiler:
  * Incorrect inlined lexical scopes as Clang allocates them
    at the scope of the enclosing function.
  * An assertion produced when in the inlining chain one of
    the original scopes has not been created.

- Fix compilation errors after upstream change:
  Remove DITypeRefArray in favour of DITypeArray
  https://github.com/llvm/llvm-project/pull/177066

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Add support for LLVM IR format.

As the IR format uses the DWARF symbolic constants,
the setting of the source language must use the DWARF
language definitions.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Add support for LLVM IR format.

If the IR parsing fails, print an explanatory error message.

jalopezg-git commented on llvm/lib/Transforms/Utils/DebugSSAUpdater.cpp:
Nit: should we remove this `assert()` (or at least remove the now-duplicated string literal)?

```suggestion
assert(NumRecordsFound == 1);
```

CarlosAlbertoEnciso replied:
Good point. Removed the `assert`.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/LVReaderHandler.cpp:
Nit: is it LLVM IR plain-text representation or bitcode form?  Here it says 'LLVM IR textual representation'; however, in [`handleObject()`, line 273](https://github.com/llvm/llvm-project/pull/135440/changes#diff-d25fa7795c3538b036476c68da7e5ff1312dcfc16bc0c2ab2d7979199ba00a98R273), it says 'Bitcode IR'.  According to the test cases, I think it is actually both (`.ll`, and `.bc`); what about saying just 'LLVM IR'?

CarlosAlbertoEnciso replied:
You are correct. At that point we are assuming both formats.
Changed to `LLVM IR`

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/LVReaderHandler.cpp:
I think I am missing something 🤔 ... this `dyn_cast<MemoryBufferRef>(&Buffer)` should always succeed, no? (and if so, the `return` in [line 302](https://github.com/llvm/llvm-project/pull/135440/changes#diff-d25fa7795c3538b036476c68da7e5ff1312dcfc16bc0c2ab2d7979199ba00a98R302) is unreachable).

CarlosAlbertoEnciso replied:
Good catch. Removed the unreachable code.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Core/LVReader.cpp:
```suggestion
  std::function<void(LVScope *Parent)> TraverseScope = [&](LVScope *Parent) {
```

CarlosAlbertoEnciso replied:
Removed extra space.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Core/LVReader.cpp:
Why was `static`-linkage changed?  This should probably stay 🤔 (?)

CarlosAlbertoEnciso replied:
You are correct. Added the `static`.

jalopezg-git commented on llvm/test/tools/llvm-debuginfo-analyzer/IR/Inputs/unit-1.ll:
This, `llvm/test/tools/llvm-debuginfo-analyzer/IR/Inputs/unit-2.ll`, and `llvm/test/tools/llvm-debuginfo-analyzer/IR/Inputs/unit-3.ll`, look similar except for the name of the `_Z5foo_1i` function.

Do they aim at testing different things?  Other than that they look duplicate -- if they are not, probably we could give these files more descriptive names in order to differentiate them.

CarlosAlbertoEnciso replied:
The files `unit-1`, `unit-2` and `unit-3` have the same functionality; the only differences are the parameter and variable names. They represent the compile units that along with `unit` will be combined IR.
Added the prefix `compile-` to all of them.

jalopezg-git commented on llvm/test/tools/llvm-debuginfo-analyzer/IR/05-ir-incorrect-lexical-scope-variable.test:
I could not observe this below.  Was this an issue that is now fixed?  Should we update this comment now?

CarlosAlbertoEnciso replied:
It seems that the original issue is now fixed.
I used the same command line as the one described in the referenced PR.
Now Clang generates correct scopes for `Var_1` and `Var_2`. Same as `GCC`.
Updated the comment.

jalopezg-git commented on llvm/unittests/DebugInfo/LogicalView/IRReaderTest.cpp:
These function seem to be shared at least with [`DWARFReaderTest.cpp`](https://github.com/llvm/llvm-project/blob/main/llvm/unittests/DebugInfo/LogicalView/DWARFReaderTest.cpp).

Perhaps, we could put them in a single place where they can be reused.  Because this is test code, this in not mandatory now and can be done as a follow-up.

CarlosAlbertoEnciso replied:
You are correct. That functionality is the same for all readers test (IR, DWARF and CodeView).
It can be done in a follow-up patch and create a shared function.

jalopezg-git commented on llvm/unittests/DebugInfo/LogicalView/IRReaderTest.cpp:
Nit: I am curious as to why is this required 🤔 ?  If it is not needed, we could also remove the `#include "llvm/Support/COM.h"`.

CarlosAlbertoEnciso replied:
It is not required. Removed both.

jalopezg-git commented on llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst:
I wonder whether we need such section 🤔 .  This seems very similar to the other 'SELECT LOGICAL ELEMENTS` sections (see, e.g. [here](https://github.com/CarlosAlbertoEnciso/llvm-project/blob/5a229edaf97c9bfe07f8372a2ac9f2c757a4153f/llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst#select-logical-elements-1)).

Having very similar sections adds little information, and makes the manual needlessly long.  Should we remove it?  WDYT?

This may be done as follow-up to improve the documentation though.

CarlosAlbertoEnciso replied:
That is a good point. The section does not add any extra information.
Removed the section.

jalopezg-git commented on llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst:
Ditto.  This looks very similar to [this other](https://github.com/CarlosAlbertoEnciso/llvm-project/blob/5a229edaf97c9bfe07f8372a2ac9f2c757a4153f/llvm/docs/CommandGuide/llvm-debuginfo-analyzer.rst#comparison-mode) section.

Again, this may be done as a follow-up PR 🙂 .

CarlosAlbertoEnciso replied:
That is a good point. The section does not add any extra information.
Removed the section.

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
IIUC, these two are mutually-exclusive; we could perhaps consider the use of `std::variant` here.  WDYT?

CarlosAlbertoEnciso replied:
Good catch. Replacing it with:
`
PointerUnion<object::IRObjectFile *, MemoryBufferRef *> InputFile;
`

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
A common style guideline is to declare all the data members s.t. they are closer to one another in the source file.  This allows to re-order them according to different criterias, e.g. reducing the amount of padding bytes, or increasing readability.

I suggest moving this to below [line 64](https://github.com/llvm/llvm-project/pull/135440/changes#diff-a605aa606ce7d0b18d33991b4aacbd646fbdad4edd329675f1d547b6f1b40ed9R64).

CarlosAlbertoEnciso replied:
Follow your suggested moves to comply with the style guideline.

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
This should be `static` and `constexpr`.  Because it becomes a compile-time constant, this should also be renamed to be `OFFSET_INCREASE`.
```suggestion
  static constexpr unsigned OFFSET_INCREASE = 4;
```

CarlosAlbertoEnciso replied:
Good suggestion.

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
Ditto (here and all occurrences below), for grouping data members together.

CarlosAlbertoEnciso replied:
Follow your suggested moves to comply with the style guideline.

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
Let's just use [`try_emplace()`](https://en.cppreference.com/cpp/container/map/try_emplace) (which avoids an additional lookup), i.e.
```suggestion
    auto [Iter, Inserted] = IndexFiles.try_emplace(CompileUnit, FileIndex);
    if (!Inserted)
      Iter->second = FileIndex;
```

CarlosAlbertoEnciso replied:
Great suggestion. Replaced.

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
Keeping the order of elements does not seem to be important here; thus `std::unordered_map` could give us average constant time lookups here.

This would apply also to the `std::map`s below (to those where order is not important).

CarlosAlbertoEnciso replied:
Apply the change to: `LVIndexFiles`, `LVCompileUnitFiles`, `LVMDObjects`

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
Maybe just
```suggestion
    LVIndexFiles::iterator Iter = IndexFiles.find(CompileUnit);
    return Iter != IndexFiles.end() ? Iter->second : 0;
```
?

CarlosAlbertoEnciso replied:
Great suggestion. Replaced.

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
Is this correct?  I guess you meant
```suggestion
  LVLine *getLineForSeenMD(const MDNode *MD) const {
    return static_cast<LVLine *>(getElementForSeenMD(MD));
  }
```
, no?

CarlosAlbertoEnciso replied:
You are correct. Good catch.

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
This is not really LVBinaryType::ELF (nor COFF).  I guess we should register a new enumerator in LVBinaryType for this.

This should be fixed also below ([line 274](https://github.com/llvm/llvm-project/pull/135440/changes#diff-a605aa606ce7d0b18d33991b4aacbd646fbdad4edd329675f1d547b6f1b40ed9R274)).

CarlosAlbertoEnciso replied:
I see you point.

During the traversal of the debug information, if the `--internal=tag` option was specified, the `LVReader::createElement()` function collects unsupported DWARF tags and print them in `LVScopeCompileUnit::printWarnings` only for readers marked with the enumerator `LVBinaryType::ELF`.

I would suggest in a follow up patch, to rename them to something like:
enum class LVDebugInfoType { NONE, DWARD, CODEVIEW };

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVIRReader.h:
I dislike this; as it may expand to ill-formed code, e.g. `this->printAllInstructions();` would result in `this->(void *)0`.  Consider the following suggestion instead:
```suggestion
  void printAllInstructions(BasicBlock *) {};
```

Alternatively, we could edit all call sites to `#ifdef` the call expression instead.

CarlosAlbertoEnciso replied:
Taking the first suggestion.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
Ditto here and below (for changing to `std::unordered_map`; is that worth it?

CarlosAlbertoEnciso replied:
I have replaced all instances of `std::map` with the suggested `std::unordered_map`.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
Isn't all this equivalent to just the following?
```suggestion
  auto [It, _] = InlinedList.try_emplace(Origin, LVList{});
  LVList &List = It->second;
  List.push_back(Inlined);
```

I.e. trying to create an empty `LVList` if there is none associated to `Origin`.  Otherwise, just add `Inlined` to the already-existing (or newly-created) list,

CarlosAlbertoEnciso replied:
Great suggestion. Changed.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
Nit: small naming change for increased readability.
```suggestion
  static LVList EmptyList;
  auto It = InlinedList.find(Origin);
  return (It == InlinedList.end()) ? EmptyList : It->second;
```

CarlosAlbertoEnciso replied:
Replaced.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```suggestion
  constexpr unsigned LEN = 17;
```

CarlosAlbertoEnciso replied:
Good suggestion. Replaced.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
I assume `LVIRReader::getIndexType()` is not meant to be called from multiple threads.  Otherwise, there is a data race here.

If this is never entered from multiple threads concurrently, consider at least adding a comment.  If it is, thread-safety needs to be ensured.

(This applies also below, where the same pattern is used; I don't think this is entered from multiple threads, but...)

CarlosAlbertoEnciso replied:
Added the comment:
`// This function is not meant to be called from multiple threads.
`

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
This could be just handled by the `return` statement in [line 653](https://github.com/llvm/llvm-project/pull/135440/changes#diff-40f44d8cbeecda4df9d15bc7f5dc934ff5f13115a051b12f49a40fab1379b7d3R653) -- don't forget to update the `assert()` that precedes it.

CarlosAlbertoEnciso replied:
Nice suggestion. Changed.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
Nit: it was curious to see the `DT` and `Aggregate` arguments swapped in this case.  Does that make sense?  Otherwise, consider changing `getOrCreateType()`'s prototype to look the same as the other two functions.

CarlosAlbertoEnciso replied:
I see your point about consistency.
Changed the `getOrCreateType` prototype.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
Doesn't this skip `Args[0]`?  Is this (for some reason) intentional?

CarlosAlbertoEnciso replied:
`Args[0]` is the return type. As we are building the subprogram arguments, we skip that entry.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
Ditto.

CarlosAlbertoEnciso replied:
`Args[0]` is the return type. As we are building the subprogram arguments, we skip that entry.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
Perhaps just initialize `Count` to `0` in [line 1525](https://github.com/llvm/llvm-project/pull/135440/changes#diff-40f44d8cbeecda4df9d15bc7f5dc934ff5f13115a051b12f49a40fab1379b7d3R1525), as it is overwritten above if needed (?).

CarlosAlbertoEnciso replied:
You are right. Done.

jalopezg-git replied:
The `if` statement in [line 1541](https://github.com/llvm/llvm-project/pull/135440/changes#diff-40f44d8cbeecda4df9d15bc7f5dc934ff5f13115a051b12f49a40fab1379b7d3R1541) can now be removed (as `Count` is directly initialized to 0.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
Seems unused; remove.
```suggestion
```

Also: if we are not handling a specific case here, consider adding an `assert()` (or similar).

CarlosAlbertoEnciso replied:
Good catch; instead of generating an assertion, we can add a value that describe the special case.
```
      else if (mdconst::dyn_extract<GlobalValue>(Value)) {
        // We cannot describe the location of dllimport'd entities: the
        // computation of their address requires loads from the IAT.
        Parameter->setValue("Unable to describe global value");
      } else ...
```

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
Nit: is `Ty->getSizeInBits()` always an integral multiple of `CHAR_BIT`?  If not, should it be rounded up?

(I guess it doesn't matter, as it seems to be only used in [line 2108](https://github.com/llvm/llvm-project/pull/135440/changes#diff-40f44d8cbeecda4df9d15bc7f5dc934ff5f13115a051b12f49a40fab1379b7d3R2108), for debug output).

CarlosAlbertoEnciso replied:
You are right; the `Ty` and `Size` are only used for debug output; moving them inside the `LLVM_DEBUG` block.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
This looks needlessly complex; `CU->getSourceLanguage()` already returns a `LVSourceLanguage` whose pointer can be passed to `setDefaultLowerBound()`.  Or am I missing something 🤔?
```suggestion
    const LVSourceLanguage SL = CU->getSourceLanguage();
```

CarlosAlbertoEnciso replied:
I think the confusion here is because the similarity in the names.

`CU->getSourceLanguage()` returns a `DISourceLanguageName`
and
`setDefaultLowerBound` expects a `LVSourceLanguage`.

The extra step uses `LanguageName` to get the required `LVSourceLanguage`.

jalopezg-git replied:
Ah, sorry for the confusion!  Closing this thread :+1:.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```suggestion
// At this point, all scopes for the compile unit have been created.
```

CarlosAlbertoEnciso replied:
Good catch. Replaced.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```suggestion
  // Traverse the scopes tree and collect those lexical blocks that do not
```

CarlosAlbertoEnciso replied:
Good catch. Replaced.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
Consider using [`llvm::scope_exit`](https://llvm.org/doxygen/classllvm_1_1scope__exit.html) to call `options().resetInternalID()` instead.  Its destructor ensures that the value is restored on function exit (and also results in more readable / maintainable code).

The `llvm::scope_exit` object should be constructed before the [`options().setInternalID()`](https://github.com/llvm/llvm-project/pull/135440/changes#diff-40f44d8cbeecda4df9d15bc7f5dc934ff5f13115a051b12f49a40fab1379b7d3R2445) call above.

CarlosAlbertoEnciso replied:
Nice suggestion; changed to use `llvm::scope_exit`.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```suggestion
  auto TraverseChildren = [&](LVScope *Parent) {
```

CarlosAlbertoEnciso replied:
Nice suggestion.
However for the current used C++ compiler, that syntax with `recursive lambdas`, produces the following compilation error:

Visual Studio 2019 (Windows):
`'TraverseChildren': cannot be used before it is initialized`

G++ 12 (Linux)
`error: use of ‘TraverseChildren’ before deduction of ‘auto’`

It seems that feature is C++23.

Keeping the current syntax.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp:
```suggestion
  auto Traverse = [&](LVScope *Current) {
```

CarlosAlbertoEnciso replied:
See previous comment.

jalopezg-git reviewed (commented):
Thank you, @CarlosAlbertoEnciso!  I did another round of review -- find below some more comments / suggestions.

I would appreciate if some other contributors (that are familiar w/ LLVM IR DebugInfo representation) could take another look at `llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp` -- @jmorse, are you up for it?.

Other than that, I would be happy to approve if nobody has any more feedback / suggestions / complaints.

CarlosAlbertoEnciso commented:
@jalopezg-git Thanks very much for your review and comments.
I have addressed all of them, with 2 intended follow-up patches.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Add support for LLVM IR format.

- Addressed the upstream review.
- Renamed the files related to the combined IR test:
    unit.ll, unit-1, unit-2, unit-3

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Add support for LLVM IR format.

Fix issues reported by Clang format.

jalopezg-git commented on llvm/lib/DebugInfo/LogicalView/LVReaderHandler.cpp:
Related to a [previous comment](https://github.com/llvm/llvm-project/pull/135440#discussion_r3163476944): If the `llvm::dyn_cast<MemoryBufferRef>(&Buffer)` always succeeds, it probably can be changed to `llvm::cast` instead.

jalopezg-git reviewed (approved):
Hey @CarlosAlbertoEnciso!  Taking into account the last batch of comments, I did an incremental review over the last changes.

I added a couple of minor comments, but this seems to be in pretty good shape now 👍 ; thus, approving!
I requested @jmorse review, but if he (or anybody else) have no further comments, I think this work can be merged.  Good job!

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Add support for LLVM IR format.

- Addressed the upstream review.

CarlosAlbertoEnciso commented:
@jalopezg-git Thanks very much for the approval.
If no one has anything else to add, I will merge the patch.

CarlosAlbertoEnciso commented:
Due to link issues with some buildbots, reverting the change.
https://github.com/llvm/llvm-project/pull/199890
